exchanges/orderbook/depth.go

```diff
@@ -46,7 +46,7 @@ type Depth struct {
 	// validationError defines current book state and why it was invalidated.
 	validationError error
 
-	m sync.Mutex
+	m sync.RWMutex
```

Comment:
There is a good amount of writes incoming from a websocket connection on this mutex, this adds an addition ~2ns per operation than a normal lock. Change not required though. As this stops the raciness and I haven't as yet had any issues in any real world trading. 

What I would probably do in the future if it does become a problem is revert it back to a normal lock and shift the stream/buffer package directly into the orderbook package where we can lock it out once and directly check the variables as we go.

Reply:
For me, this isn't about raciness, but concurrency.
We can't control what users are doing with the methods available to them.
A RWMutex is the best way I can think of to protect write performance in the face of an unpredictable number of reads.
Even if we shift everything to one package, the previous implementations of an OB I've written have still needed a RWMutex over a straight mutex. Where I've tried to do away with the mutex by using a single select, I've found that it stopped-the-world for too long in several situations.
Maybe I'm missing your point.

Reply:
I don't think you are missing my point, we both want the same goal. What your saying makes sense for a read heavy load but the most important thing is write performance and I guess further testing can be done. 

> We can't control what users are doing with the methods available to them.
👻 They are dead to me. I have external operations that only use isolated wrapper functionality. If they are being silly billies then that's on them. 

> A RWMutex is the best way I can think of to protect write performance in the face of an unpredictable number of reads.
You're probably correct on this assumption. 

> Even if we shift everything to one package, the previous implementations of an OB I've written have still needed a RWMutex over a straight mutex. Where I've tried to do away with the mutex by using a single select, I've found that it stopped-the-world for too long in several situations.
What use case was this? Like is this for a market maker bot or similar system?